File: cjeongmin/실패율.js
const upperBound = (arr) => (x) => {
    let left = 0;
    let right = arr.length;
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] <= x) left = mid + 1;
        else right = mid;
    }
    return left;
};

function solution(N, stages) {
    stages.sort((a, b) => a - b);
    const upper = upperBound(stages);
    let curr = 0;
    return Array.from({ length: N }, (_, i) => i + 1)
        .map((stage) => [upper(stage), stage])
        .map((v) => {
            const x = (v[0] - curr) / (stages.length - curr);
            curr = v[0];
            return [x, v[1]];
        })
        .sort((a, b) => b[0] - a[0])
        .map((v) => v[1]);
}

File: jiho/실패율.py
def solution(N, stages):
    counter = [0 for i in range(N + 2)]
    for stage in stages:
        counter[stage]+=1
    
    fail = []
    total = len(stages)
    for i, c in enumerate(counter[1:-1]):
        fail.append((i+1, 0 if total == 0 else c/total))
        total -= c
    
    fail.sort(key= lambda f: (-f[1], f[0]))
    return [idx for (idx, _) in fail]